codegen/templates/tchannel_client.tmpl

```diff
@@ -0,0 +1,231 @@
+	"context"
+	"errors"
+	"strconv"
+	"time"
+
+	"github.com/uber/zanzibar/runtime"
+
+	{{range $idx, $pkg := .IncludedPackages -}}
+	{{$pkg.AliasName}} "{{$pkg.PackageName}}"
+	{{end}}
+)
+
+
+// Interfaces for the service and client for the services defined in the IDL.
+{{/* TODO: case */}}
+{{$clientID := .ClientID -}}
+{{$clientName := title .ClientID | printf "%sClient" -}}
+{{range $svc := .Services}}
+// TChan{{title $clientID}} is the interface that defines the server handler and client interface.
+type TChan{{title $clientID}} interface {
+	{{range .Methods}}
+	    {{if and (eq .RequestType "") (eq .ResponseType "")}}
+		    {{- .Name}}(
+		        ctx context.Context,
+		        reqHeaders map[string]string,
+		    ) (
+		        map[string]string,
+		        error,
+		    )
+	    {{else if eq .RequestType ""}}
+		    {{- .Name}}(
+		        ctx context.Context,
+		        reqHeaders map[string]string,
+		    ) (
+		        *{{.ResponseType}},
+		        map[string]string,
+		        error,
+		    )
+	    {{else if eq .ResponseType ""}}
+		    {{- .Name}}(
+		        ctx context.Context,
+		        reqHeaders map[string]string,
+		        args *{{.RequestType}},
+		    ) (
+		        map[string]string, error,
+		    )
+		{{else}}
+		    {{- .Name}}(
+		        ctx context.Context,
+		        reqHeaders map[string]string,
+		        args *{{.RequestType}},
+		    ) (
+		        *{{.ResponseType}},
+		        map[string]string,
+		        error,
+		    )
+		{{end}}
+	{{end -}}
+}
+
+// NewClient returns a new TChannel client for service {{$clientID}}.
+func NewClient(gateway *zanzibar.Gateway) *{{$clientName}} {
+    {{- /* this is the service discovery service name */}}
+	serviceName := gateway.Config.MustGetString("clients.{{$clientID}}.serviceName")
+	sc := gateway.Channel.GetSubChannel(serviceName)
+
+	ip := gateway.Config.MustGetString("clients.{{$clientID}}.ip")
+	port := gateway.Config.MustGetInt("clients.{{$clientID}}.port")
+	sc.Peers().Add(ip + ":" + strconv.Itoa(int(port)))
+
+	{{/* TODO: (lu) maybe set these at per method level */ -}}
+	timeout := time.Millisecond * time.Duration(
+	    gateway.Config.MustGetInt("clients.{{$clientID}}.timeout"),
+	)
+	timeoutPerAttempt := time.Millisecond * time.Duration(
+	    gateway.Config.MustGetInt("clients.{{$clientID}}.timeoutPerAttempt"),
+	)
+
+	client := zanzibar.NewTChannelClient(gateway.Channel,
+		&zanzibar.TChannelClientOption{
+			ServiceName:       serviceName,
+			Timeout:           timeout,
+			TimeoutPerAttempt: timeoutPerAttempt,
+		},
+	)
+
+	return &{{$clientName}}{
+    	{{- /* this is the thrift service name, different from service discovery service name */}}
+		thriftService: "{{$svc.Name}}",
+		client:        client,
+	}
+}
+
+// {{$clientName}} is the TChannel client for downstream service.
+type {{$clientName}} struct {
+	thriftService string
+	client        zanzibar.TChanClient
+}
+
+{{range .Methods}}
+    {{- if and (eq .RequestType "") (eq .ResponseType "")}}
```

Comment:
fix in https://github.com/uber/zanzibar/pull/57/commits/e98eab15ecda1a3c5b792f1876335468c4fd678c